File: attentive-android-sdk/src/test/java/com/attentive/androidsdk/AttentiveEventTrackerTest.java
package com.attentive.androidsdk;

import static junit.framework.TestCase.assertEquals;
import static org.junit.Assert.assertThrows;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import com.attentive.androidsdk.events.Event;
import java.lang.reflect.Field;
import org.junit.Before;
import org.junit.Test;

public class AttentiveEventTrackerTest {
    private static final String DOMAIN = "someDomainValue";
    private static final UserIdentifiers USER_IDENTIFIERS = new UserIdentifiers.Builder().build();

    private AttentiveConfig config;
    private AttentiveApi attentiveApi;

    @Before
    public void setup() throws NoSuchFieldException, IllegalAccessException {
        config = mock(AttentiveConfig.class);
        attentiveApi = mock(AttentiveApi.class);
        doReturn(attentiveApi).when(config).getAttentiveApi();
        doReturn(DOMAIN).when(config).getDomain();
        doReturn(USER_IDENTIFIERS).when(config).getUserIdentifiers();

        resetSingleton();
    }

    public void resetSingleton() throws SecurityException, NoSuchFieldException, IllegalArgumentException, IllegalAccessException {
        Field instance = AttentiveEventTracker.class.getDeclaredField("INSTANCE");
        instance.setAccessible(true);
        instance.set(null, null);
    }

    @Test
    public void getInstance_doesNotThrow() {
        // Verify does not throw
        AttentiveEventTracker.getInstance();
    }

    @Test
    public void getInstance_calledTwice_returnsSameInstance() {
        assertEquals(AttentiveEventTracker.getInstance(), AttentiveEventTracker.getInstance());
    }

    @Test
    public void initialize_nullConfig_throws() {
        assertThrows(IllegalArgumentException.class, () -> AttentiveEventTracker.getInstance().initialize(null));
    }

    @Test
    public void initialize_validConfig_success() {
        AttentiveEventTracker.getInstance().initialize(config);
    }

    @Test
    public void initialize_calledTwice_doesNotThrow() {
        AttentiveEventTracker.getInstance().initialize(config);
    }

    @Test
    public void recordEvent_nullEvent_throws() {
        AttentiveEventTracker.getInstance().initialize(config);
        assertThrows(IllegalArgumentException.class, () -> AttentiveEventTracker.getInstance().recordEvent(null));
    }

    @Test
    public void recordEvent_validEvent_sendsToApi() {
        // Arrange
        AttentiveEventTracker.getInstance().initialize(config);
        Event eventToSend = mock(Event.class);

        // Act
        AttentiveEventTracker.getInstance().recordEvent(eventToSend);

        // Assert
        verify(attentiveApi).sendEvent(eventToSend, USER_IDENTIFIERS, DOMAIN);
    }
}

File: attentive-android-sdk/src/test/java/com/attentive/androidsdk/AttentiveEventTrackerTest.kt
package com.attentive.androidsdk

import android.app.Application
import android.os.Looper
import androidx.lifecycle.Lifecycle
import com.attentive.androidsdk.AttentiveApi
import com.attentive.androidsdk.AttentiveConfig
import com.attentive.androidsdk.AttentiveEventTracker
import com.attentive.androidsdk.events.Event
import com.attentive.androidsdk.tracking.AppLaunchTracker
import junit.framework.TestCase
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.kotlin.spy
import org.mockito.kotlin.whenever
import kotlin.jvm.java

class AttentiveEventTrackerTest {
    lateinit var config: AttentiveConfig
    private var attentiveApi: AttentiveApi? = null
    private var launchTracker: AppLaunchTracker? = null
    val mockLifecycle = Mockito.mock(Lifecycle::class.java)
    val instance = Mockito.spy(AttentiveEventTracker::class.java)


    @Before
    @Throws(NoSuchFieldException::class, IllegalAccessException::class)
    fun setup() {
        config = Mockito.mock(AttentiveConfig::class.java)
        attentiveApi = Mockito.mock(AttentiveApi::class.java)
        launchTracker = Mockito.mock(AppLaunchTracker::class.java)
        Mockito.doReturn(attentiveApi).`when`(config)?.attentiveApi
        Mockito.doReturn(DOMAIN).`when`(config)?.domain
        Mockito.doReturn(USER_IDENTIFIERS).`when`(config)?.userIdentifiers
        Mockito.doReturn(mockLifecycle).`when`(launchTracker)?.lifecycle
        Mockito.doReturn(launchTracker).whenever(instance).launchTracker
        whenever(config.applicationContext).thenReturn(Mockito.mock(Application::class.java))
    }
//
//    @Test
//    fun initialize_validConfig_success() {
//        instance.initialize(config)
//    }
//
//    @Test
//    fun initialize_calledTwice_doesNotThrow() {
//        instance.initialize(config)
//    }
//
//
//    @Test
//    fun recordEvent_validEvent_sendsToApi() {
//        // Arrange
//        instance.initialize(config)
//        val eventToSend = Mockito.mock(
//            Event::class.java
//        )
//
//        // Act
//        instance.recordEvent(eventToSend)
//
//        // Assert
//        Mockito.verify(attentiveApi)?.sendEvent(eventToSend, USER_IDENTIFIERS, DOMAIN)
//    }

    companion object {
        private const val DOMAIN = "someDomainValue"
        private val USER_IDENTIFIERS = UserIdentifiers.Builder().build()
    }
}